sphinx/directives/__init__.py

```diff
@@ -203,8 +205,30 @@ def run(self) -> List[Node]:
         DocFieldTransformer(self).transform_all(contentnode)
         self.env.temp_data['object'] = None
         self.after_content()
+
+        if node['hidden']:
+            node = self.replace_node_with_target(node)
+            self.set_source_info(node)
+
         return [self.indexnode, node]
 
+    # TODO: This method does not need access to self, make a (non-class) function out of it?
+    def replace_node_with_target(self, node: nodes.Node) -> nodes.target:
+
+        def collect_ids(node: nodes.Node) -> List[str]:
+            if isinstance(node, nodes.Element):
+                ids: List[str] = node.get('ids', [])
+                for c in node.children:
+                    ids.extend(collect_ids(c))
+                return ids
+            else:
+                return []
```

Comment:
Does this function need to be defined as a closure? If not please factor out for speed.